### PR TITLE
Martin/jib 196 indicate which message is most recent

### PR DESCRIPTION
**Pull request by MartinJibber**
https://user-images.githubusercontent.com/88675954/140436092-68d2d90a-1bcf-48d1-98ee-506ae079bf63.mp4

### CODE REVIEW

File: Shared/Resources/Assets.xcassets/Colors/TEXTCOLOR.colorset/Contents.json [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
     }
-  ]
-}
\ No newline at end of file
+  ],
+  "info" : {
+    "author" : "xcode",
+    "version" : 1
+  }
+}
```
Comment (MartinJibber) [inline Shared/Resources/Assets.xcassets/Colors/TEXTCOLOR.colorset/Contents.json]: Changing text color to black.

File: iOS/Flows/Conversation/Cells/ConversationMessageCell.swift [hunk 2]
```diff
@@ -18,9 +18,9 @@ class ConversationMessageCell: UICollectionViewCell {
     var handleTappedMessage: ((Messageable) -> Void)?
     var handleDeleteMessage: ((Messageable) -> Void)?
 
-    private let collectionView: UICollectionView = {
-        let layout = ConversationMessageCellLayout()
-        let cv = UICollectionView(frame: .zero, collectionViewLayout: layout)
+    private let collectionLayout = ConversationMessageCellLayout()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCell.swift]: The cell layout needs to have a message controller assigned to it so I'm making it a property.

File: Shared/Models/Message Model/Messageable.swift [hunk 3]
```diff
@@ -32,7 +33,7 @@ protocol Messageable {
     var color: Color { get }
     var kind: MessageKind { get }
     var isDeleted: Bool { get }
-    var conversationId: String { get }
+    var recentReplies: [Messageable] { get }
```
Comment (MartinJibber) [inline Shared/Models/Message Model/Messageable.swift:36]: Adding this to messageable so we can determine if a cell is the most recently sent one.

File: Shared/Resources/Assets.xcassets/Colors/TEXTCOLOR.colorset/Contents.json (hunk 1 above)
Comment (MartinJibber) [inline Shared/Resources/Assets.xcassets/Colors/TEXTCOLOR.colorset/Contents.json]: Making the text color black.

File: iOS/Flows/Conversation/Cells/ConversationMessageCell.swift [hunk 4]
```diff
@@ -12,15 +12,15 @@ import UIKit
...
+class ConversationMessageCell: UICollectionViewCell, ConversationMessageCellLayoutDelegate {
 
     // Interaction handling
     var handleTappedMessage: ((Messageable) -> Void)?
     var handleDeleteMessage: ((Messageable) -> Void)?
 
-    private let collectionView: UICollectionView = {
-        let layout = ConversationMessageCellLayout()
-        let cv = UICollectionView(frame: .zero, collectionViewLayout: layout)
+    private lazy var collectionLayout = ConversationMessageCellLayout(messageDelegate: self)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCell.swift:21]: The layout needs access to know which message is the latest one, so we pass in ourselves as a message delegate. More on this later.

File: iOS/Flows/Conversation/Cells/ConversationMessageCell.swift [hunk 5]
```diff
@@ -29,7 +29,7 @@ class ConversationMessageCell: UICollectionViewCell {
     private var state: ConversationUIState = .read
 
     /// The parent message of this thread.
-    private var message: Messageable?
+    var message: Messageable?
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCell.swift:32]: The delegate protocol requires this be public.

File: iOS/Flows/Conversation/Cells/ConversationMessageCellDatasource.swift [hunk 6]
```diff
@@ -116,22 +116,15 @@ extension ConversationMessageCellDataSource {
...
             let dataSource = item.dataSource
-            // The higher the cells index, the closer it is to the front of the message stack.
-            let stackIndex = dataSource.getStackIndex(forIndexPath: indexPath)
-
-            let backgroundColor: UIColor
-            if message.isFromCurrentUser {
-                backgroundColor = .gray
-            } else {
-                backgroundColor = .darkGray
-            }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCellDatasource.swift:127]: Unused.

File: iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift [hunk 7]
```diff
@@ -8,20 +8,26 @@
...
 
+    unowned let messageDelegate: ConversationMessageCellLayoutDelegate
+
     override class var layoutAttributesClass: AnyClass {
         return ConversationMessageCellLayoutAttributes.self
     }
 
-    override init() {
+    init(messageDelegate: ConversationMessageCellLayoutDelegate) {
+        self.messageDelegate = messageDelegate
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift:24]: As mentioned above, the layout class has a delegate to get the parent message of a thread.

File: iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift [hunk 8]
```diff
@@ -56,26 +62,41 @@ class ConversationMessageCellLayout: UICollectionViewFlowLayout {
...
         backgroundColor = backgroundColor.color(withBrightness: backgroundBrightness)
 
         attributes.backgroundColor = backgroundColor
 
-        // Only show text for the front most item in each section.
-        attributes.shouldShowText = stackIndex == 0
+        var isMostRecentMessageFromUser = false
+        if let mostRecentMessage = messageDelegate.message?.recentReplies.first {
+            isMostRecentMessageFromUser = mostRecentMessage.isFromCurrentUser
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift:78]: Here's where we get the message from the delegate.

File: iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift [hunk 9]
```diff
@@ -56,26 +62,41 @@ class ConversationMessageCellLayout: UICollectionViewFlowLayout {
...
 
         if indexPath.section == 1 {
             // The second section should have a tail on its last item
             attributes.shouldShowTail = stackIndex == 0
             attributes.bubbleTailOrientation = .down
+
+            // If the most recent message is from the current user, then highlight it with a white background.
+            if isMostRecentMessageFromUser && stackIndex == 0 {
+                attributes.backgroundColor = .white
+            }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift:99]: Make the top cell white based upon who sent the message.

File: iOS/Flows/Conversation/ConversationViewController.swift [hunk 10]
```diff
@@ -26,7 +26,7 @@ class ConversationViewController: FullScreenViewController,
     /// Denotes where a message should be dragged and dropped to send.
     private let sendMessageOverlay = ConversationSendOverlayView()
     
-    private let blurView = UIVisualEffectView(effect: UIBlurEffect(style: .dark))
+    private let blurView = UIVisualEffectView(effect: UIBlurEffect(style: .light))
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewController.swift:29]: Making the conversation background lighter.

File: iOS/Flows/Conversation/Threads/Conversation Collection Views/Cells/Message/MessageTextView.swift [hunk 11]
```diff
@@ -25,13 +25,13 @@ class MessageTextView: TextView {
     }
 
     func set(text: Localized, messageContext: MessageContext) {
-        let textColor: Color = messageContext == .status ? .darkGray : .white
+        let textColor: Color = messageContext == .status ? .darkGray : .textColor
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Threads/Conversation Collection Views/Cells/Message/MessageTextView.swift:28]: Changed text color to black and used it here.

**Review by Benjamindodgson — APPROVED** [review]

File: iOS/Flows/Conversation/Threads/Conversation Collection Views/Cells/Message/MessageTextView.swift (hunk 11 above)
Comment (Benjamindodgson) [reply iOS/Flows/Conversation/Threads/Conversation Collection Views/Cells/Message/MessageTextView.swift:28]: nice

**Pull request merged**